db/blob/blob_source.cc

```diff
@@ -0,0 +1,238 @@
+//  Copyright (c) Meta Platforms, Inc. and affiliates.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+
+#include "db/blob/blob_source.h"
+
+#include <cassert>
+#include <string>
+
+#include "db/blob/blob_file_reader.h"
+#include "file/file_prefetch_buffer.h"
+#include "options/cf_options.h"
+
+namespace ROCKSDB_NAMESPACE {
+
+BlobSource::BlobSource(Cache* cache, const ImmutableOptions* immutable_options,
+                       const FileOptions* file_options,
+                       const std::string& db_id,
+                       const std::string& db_session_id,
+                       uint32_t column_family_id,
+                       HistogramImpl* blob_file_read_hist,
+                       const std::shared_ptr<IOTracer>& io_tracer)
+    : db_id_(db_id),
+      db_session_id_(db_session_id),
+      statistics_(immutable_options->statistics.get()),
+      blob_file_cache_(new BlobFileCache(cache, immutable_options, file_options,
+                                         column_family_id, blob_file_read_hist,
+                                         io_tracer)),
+      blob_cache_(immutable_options->blob_cache) {}
+
+BlobSource::~BlobSource() = default;
+
+Status BlobSource::MaybeReadBlobFromCache(
+    FilePrefetchBuffer* prefetch_buffer, const ReadOptions& read_options,
+    const CacheKey& cache_key, uint64_t offset,
+    CachableEntry<Slice>* blob_entry) const {
+  assert(blob_entry != nullptr);
+  assert(blob_entry->IsEmpty());
+
+  // First, try to get the blob from the cache
+  //
+  // If either blob cache is enabled, we'll try to read from it.
```

Comment:
Nit: is `either` here a typo?

Reply:
Thanks. I should remove `either` since I removed the comments for secondary cache.